Code/Source/Wrappers/Texture.cpp

```diff
@@ -0,0 +1,210 @@
+        static const AZ::Name albedoColorName = AZ::Name::FromStringLiteral("baseColor.color", AZ::Interface<AZ::NameDictionary>::Get());
+
+        Texture imageTexture = CreateInvalid();
+
+        const AZ::Data::AssetId& id = materialAsset.GetId();
+        if (!materialAsset.IsReady())
+        {
+            AZ_Warning(TraceWindow.c_str(), false, "The material asset with ID: %s was not ready.", id.ToString<AZStd::string>().c_str());
+            return imageTexture;
+        }
+
+        const auto* propLayout = materialAsset->GetMaterialPropertiesLayout();
+
+        if (!propLayout)
+        {
+            AZ_Warning(
+                TraceWindow.c_str(),
+                false,
+                "Unable to access material properties layout of material asset with ID: %s.",
+                id.ToString<AZStd::string>().c_str());
+            return imageTexture;
+        }
+
+        auto& propertyValues = materialAsset->GetPropertyValues();
+
+        if (const auto albedoTexPropIdx = propLayout->FindPropertyIndex(albedoTexName); !albedoTexPropIdx.IsNull())
+        {
+            const auto albedoTexImagePropVal = propertyValues.at(albedoTexPropIdx.GetIndex());
+            const auto& imageAsset = albedoTexImagePropVal.GetValue<AZ::Data::Asset<AZ::RPI::ImageAsset>>();
+            imageTexture = CreateFromImageAsset(imageAsset.GetId());
+        }
+
+        if (imageTexture.IsValid())
+        {
+            return imageTexture;
+        }
+
+        if (const auto albedoColorPropIdx = propLayout->FindPropertyIndex(albedoColorName); !albedoColorPropIdx.IsNull())
+        {
+            imageTexture =
+                AZStd::move(CreateFromFactor(GrayFromColor(propertyValues.at(albedoColorPropIdx.GetIndex()).GetValue<AZ::Color>())));
+        }
+        else
+        {
+            AZ_Error(
+                TraceWindow.c_str(),
+                false,
+                "Unable to find specular color and texture properties of material asset with ID: %s.",
+                id.ToString<AZStd::string>().c_str());
+        }
+
+        return imageTexture;
+    }
+
+    Texture Texture::CreateFromFactor(float factor)
+    {
+        auto factor8 = aznumeric_cast<AZ::u8>(factor * 255.0f);
+        return Texture{ &factor8, 1, 1 };
+    }
+
+    float Texture::GrayFromColor(const AZ::Color& color)
+    {
+        return RedGrayMultiplier * color.GetR() + GreenGrayMultiplier * color.GetG() + BlueGrayMultiplier * color.GetB();
+    }
+
+    uint8_t Texture::Gray8FromColor(const AZ::Color& color)
+    {
+        return static_cast<uint8_t>(GrayFromColor(color) * 255.0f);
+    }
+
+    Texture Texture::CreateFromImageAsset(const AZ::Data::AssetId& imageAssetId)
+    {
+        // Made a static member to minimize reallocations, since multiple assets will be processed.
+        static AZStd::vector<uint8_t> tempRglTextureData;
```

Comment:
This may be an overoptimization? When you resize and clear, with next resize you will still probably get reallocation. And if for some reason CreateFromImageAsset will be called asynchronously, it may easily break.

Reply:
I've tested this with a closed source project and these are the results:
```bash
(CreateFromImageAsset) - Resizing tempRglTexture data. From 0 to 5064000 bytes.
(CreateFromImageAsset) - Reallocation occured.
(CreateFromImageAsset) - Resizing tempRglTexture data. From 5064000 to 81920 bytes.
(CreateFromImageAsset) - Resizing tempRglTexture data. From 81920 to 14745600 bytes.
(CreateFromImageAsset) - Reallocation occured.
(CreateFromImageAsset) - Resizing tempRglTexture data. From 14745600 to 1048576 bytes.
(CreateFromImageAsset) - Resizing tempRglTexture data. From 1048576 to 16777216 bytes.
(CreateFromImageAsset) - Reallocation occured.
(CreateFromImageAsset) - Resizing tempRglTexture data. From 16777216 to 16777216 bytes.
(CreateFromImageAsset) - Resizing tempRglTexture data. From 16777216 to 16777216 bytes.
(CreateFromImageAsset) - Resizing tempRglTexture data. From 16777216 to 16777216 bytes.
(CreateFromImageAsset) - Resizing tempRglTexture data. From 16777216 to 1048576 bytes.
(CreateFromImageAsset) - Resizing tempRglTexture data. From 1048576 to 16777216 bytes.
(CreateFromImageAsset) - Resizing tempRglTexture data. From 16777216 to 4194304 bytes.
(CreateFromImageAsset) - Resizing tempRglTexture data. From 4194304 to 16777216 bytes.
(CreateFromImageAsset) - Resizing tempRglTexture data. From 16777216 to 16777216 bytes.
(CreateFromImageAsset) - Resizing tempRglTexture data. From 16777216 to 16777216 bytes.
(CreateFromImageAsset) - Resizing tempRglTexture data. From 16777216 to 16777216 bytes.
(CreateFromImageAsset) - Resizing tempRglTexture data. From 16777216 to 811008 bytes.
(CreateFromImageAsset) - Resizing tempRglTexture data. From 811008 to 16777216 bytes.
```
I've moved the clear just before the resize and tested if the pointer changed before and after the clear and resize. From those results I would say this is a perfectly valid place for such an optimization. Additionally I don't expect this piece of code run in parallel anytime soon.